scipy/signal/tests/test_short_time_fft.py

```diff
@@ -0,0 +1,775 @@
+"""Unit tests for module `_short_time_fft`.
+
+This file's structure loosely groups the tests into the following sequential
+categories:
+
+1. Test function `_calc_dual_canonical_window`.
+2. Test for invalid parameters and exceptions in `ShortTimeFFT` (until the
+    `test_from_window` function).
+3. Test algorithmic properties of STFT/ISTFT. Some tests were ported from
+   ``test_spectral.py``.
+
+Notes
+-----
+* The coverage of `_short_time_fft` is only 99% because the exception block for
+  testing if `NDArray` exists in Numpy, i.e. the following block
+
+        try:  # Workaround needed for Numpy versions < 1.21
+            from numpy.typing import NDArray
+        except ModuleNotFoundError:
+            NDArray: Type = np.ndarray  # type: ignore
+
+  is not tested. The reason being that the author does not know hot to do it in
+  a reasonable manner within pytest. This problem will go away when Numpy
+  >= 1.21 becomes mandatory for SciPy.
+* Mypy 0.990 does interpret the line::
+
+        from scipy.stats import norm as normal_distribution
+
+  incorrectly (but the code works), hence a ``type: ignore`` was appended.
+"""
+import math
+from itertools import product
+from typing import cast, get_args, Literal
+
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose, assert_equal
+from scipy.fft import fftshift
+from scipy.stats import norm as normal_distribution  # type: ignore
+from scipy.signal import get_window, welch, stft, istft, spectrogram
+
+from scipy.signal._short_time_fft import FFT_TYP_TYPE, \
+    _calc_dual_canonical_window, ShortTimeFFT, PAD_TYPE
+from scipy.signal.windows import gaussian
+
+
+def chk_VE(match):
+    """Assert for a ValueError matching regexp `match`.
+
+    This little wrapper allows a more concise code layout. Usage::
+
+        with chk_VE("Message RegExp"):
+            test_code
+            ...
+    """
+    return pytest.raises(ValueError, match=match)
```

Comment:
I would not do this. For maintenance this complicate our task as the code does not look familiar compared to other places.